app/models/project_step.rb

```diff
@@ -135,6 +138,27 @@ def handle_original_date_logic
     end
   end
 
+  def handle_finalized_at
+    if is_finalized && ! finalized_at
+      self.finalized_at = Time.now
+    elsif ! is_finalized && finalized_at
+      self.finalized_at = nil
+    end
+  end
+
+  # Validates that a step may not be unfinalized more than 24 hours since it was previously marked
+  # as finalized.  Note, should generally be avoided by front-end logic, but guards against edge
+  # cases.
+  def unfinalize_allowed
+    if is_finalized_changed? && ! is_finalized && is_finalized_locked?
+      errors.add(:is_finalized, I18n.t("project_step.unfinalize_disallowed"))
+    end
+  end
```

Comment:
Fuzzy said he's ok with refactoring that so go for it. I'd recommend a new
method like `best_known_date` or something that does the check and returns
the most up to date value.

On Fri, Apr 29, 2016 at 9:43 AM, Joseph Elliott <notifications@github.com>
wrote:

> In app/models/project_step.rb
> <https://github.com/sassafrastech/madeline_system/pull/69#discussion_r61579270>
> :
>
> > +  def handle_finalized_at
> > +    if is_finalized && ! finalized_at
> > +      self.finalized_at = Time.now
> > +    elsif ! is_finalized && finalized_at
> > +      self.finalized_at = nil
> > +    end
> > +  end
> > +
> > +  # Validates that a step may not be unfinalized more than 24 hours since it was previously marked
> > +  # as finalized.  Note, should generally be avoided by front-end logic, but guards against edge
> > +  # cases.
> > +  def unfinalize_allowed
> > +    if is_finalized_changed? && ! is_finalized && is_finalized_locked?
> > +      errors.add(:is_finalized, I18n.t("project_step.unfinalize_disallowed"))
> > +    end
> > +  end
>
> Yeah. I've added tests for the finalize handling as well as the automatic
> original_date updates. (and I got tripped up yet one more time with the
> confusing original_date getter which masks an empty db value)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/sassafrastech/madeline_system/pull/69/files/384c30beb7e505b87ce1fa30d40b434b91106b53#r61579270>
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org